tests/php/Emitter/OutputEmitter/NodeEmitter/ApplyEmitterTest.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+declare(strict_types=1);
+
+namespace PhelTest\Emitter\OutputEmitter\NodeEmitter;
+
+use Phel\Ast\ApplyNode;
+use Phel\Ast\FnNode;
+use Phel\Ast\LiteralNode;
+use Phel\Ast\PhpVarNode;
+use Phel\Ast\TupleNode;
+use Phel\Emitter\OutputEmitter;
+use Phel\Emitter\OutputEmitter\NodeEmitter\ApplyEmitter;
+use Phel\Lang\Symbol;
+use Phel\NodeEnvironment;
+use PHPUnit\Framework\TestCase;
+
+final class ApplyEmitterTest extends TestCase
+{
+    private ApplyEmitter $applyEmitter;
+
+    public function setUp(): void
+    {
+        $this->applyEmitter = new ApplyEmitter(
+            OutputEmitter::createWithSourceMap()
+        );
+    }
+
+    public function testPhpVarNodeAndFnNodeIsInfix(): void
+    {
+        $node = new PhpVarNode(NodeEnvironment::empty(), '+');
+        $args = [
+            new TupleNode(NodeEnvironment::empty(), [
+                new LiteralNode(NodeEnvironment::empty(), 2),
+                new LiteralNode(NodeEnvironment::empty(), 3),
+                new LiteralNode(NodeEnvironment::empty(), 4),
+            ]),
+        ];
+
+        $applyNode = new ApplyNode(NodeEnvironment::empty(), $node, $args);
+        $this->applyEmitter->emit($applyNode);
+
+        $this->expectOutputString(
+            'array_reduce([...((\Phel\Lang\Tuple::createBracket(, , );) ?? [])], function($a, $b) { return ($a + $b); });'
+        );
+    }
+
+    public function testPhpVarNodeButNoInfix(): void
+    {
+        $node = new PhpVarNode(NodeEnvironment::empty(), 'str');
+        $args = [
+            new LiteralNode(NodeEnvironment::empty(), 'abc'),
+            new TupleNode(NodeEnvironment::empty(), [
+                new LiteralNode(NodeEnvironment::empty(), 'def'),
+            ]),
+        ];
+
+        $applyNode = new ApplyNode(NodeEnvironment::empty(), $node, $args);
+        $this->applyEmitter->emit($applyNode);
+
+        $this->expectOutputString('str(, ...((\Phel\Lang\Tuple::createBracket();) ?? []));');
+    }
+
+    public function testNoPhpVarNode(): void
+    {
+        $fnNode = new FnNode(
+            NodeEnvironment::empty(),
+            [Symbol::create('&'), Symbol::create('x')],
```

Comment:
You can remove the `&` Symbol here. In the full pipeline the Analyzer will remove this and sets `isVariadic` to true (see next comment).

Should look like this then:
```
[Symbol::create('x')]
```

Reply:
In your previous message you said:
> but this one is even easier (but doesn't make much sense)
> ```clojure
> (apply (fn [&]) [1])
> ```

And if I removed the `Symbol::create('&')`, the final query would be:
```clojure
(apply (fn [x]) [1])
```
Are they the same?

Reply:
Good question!

First of all, the generated code of `(fn [& x])` and `(fn [&])` is almost the same. I the second example the analyzer will automatically generated a dummy symbol. So the analyzer will transform it to something like this `(fn [& __phel_1233])`

Now let me explain the `&` Symbol. This is just a hint for the analyzer that the next argument is a variadic argument. This is equal to the `...` before a function argument in PHP. These two are equal
```php
<?php
function(...$args) {}
```

```
(fn [& args])
```

So the analyzer recognizes the `&` symbol. But for the Emitter this `&` has no meaning. The emitter just wants to know if the last argument of the function is a variadic argument (so it has to add the three dots). Therefore, the analyzer removes the `&` symbol from the argument list and sets the variadic flag to true.

Hope this makes it clear now ;)

Reply:
Thank you!